src/field_theory/splitting_field.lean

```diff
@@ -109,46 +109,226 @@ namespace splitting_field_aux
+    rat_cast := splitting_field_aux.mk n ∘ (coe : ℚ → K),
+    rat_cast_mk := have h : _ := @field.rat_cast_mk, _,
+    qsmul := (•),
+    qsmul_eq_mul' := have h : _ := @field.qsmul_eq_mul', _,
+    npow := splitting_field_aux.npow n,
+    npow_zero' := have h : _ := @comm_ring.npow_zero', _,
+    npow_succ' := have h : _ := @comm_ring.npow_succ', _,
+    zpow := splitting_field_aux.zpow n,
+    zpow_zero' := have h : _ := @field.zpow_zero', _,
+    zpow_succ' := have h : _ := @field.zpow_succ', _,
+    zpow_neg' := have h : _ := @field.zpow_neg', _},
+  all_goals {
+    unfreezingI { induction n with n ih generalizing K },
+    { apply @h K },
+    { exact @ih _ _ _ } },
+end
+
+instance inhabited {n : ℕ} {f : K[X]} :
+  inhabited (splitting_field_aux n f) := ⟨37⟩
+
+/-- The injection from the base field as a ring homomorphism. -/
+@[simps] protected def mk_hom (n : ℕ) {K : Type u} [field K] {f : K[X]} :
+  K →+* splitting_field_aux n f :=
+{ to_fun := splitting_field_aux.mk n,
+  map_one' :=
+  begin
+    unfreezingI { induction n with k hk generalizing K },
+    { simp [splitting_field_aux.mk] },
+    exact hk
+  end,
+  map_mul' :=
+  begin
+    unfreezingI { induction n with k hk generalizing K },
+    { simp [splitting_field_aux.mk] },
+    intros x y,
+    change (splitting_field_aux.mk k) ((adjoin_root.of f.factor) _) = _,
+    rw [map_mul],
+    exact hk _ _
+  end,
+  map_zero' :=
+  begin
+    unfreezingI { induction n with k hk generalizing K },
+    { simp [splitting_field_aux.mk] },
+    change (splitting_field_aux.mk k) ((adjoin_root.of f.factor) 0) = _,
+    rw [map_zero, hk],
+  end,
+  map_add' :=
+  begin
+    unfreezingI { induction n with k hk generalizing K },
+    { simp [splitting_field_aux.mk] },
+    intros x y,
+    change (splitting_field_aux.mk k) ((adjoin_root.of f.factor) _) = _,
+    rw [map_add],
+    exact hk _ _
+  end }
+
+-- Note that all fields of this algebra instance are equal to the `{nat/int/rat}_cast` and
+-- `{n/z/q}smul`; this can be done by using `mk_hom`, definitionally equal to `mk`.
+instance algebra (n : ℕ) (R : Type*) {K : Type u} [comm_semiring R] [field K]
+  [algebra R K] {f : K[X]} : algebra R (splitting_field_aux n f) :=
+{ to_fun := (splitting_field_aux.mk_hom n).comp (algebra_map R K),
+  smul := (•),
+  smul_def' :=
+  begin
+    unfreezingI { induction n with k hk generalizing K },
+    { exact algebra.smul_def },
+    exact hk
+  end,
+  commutes' := λ a b, mul_comm _ _,
+  .. (splitting_field_aux.mk_hom n).comp (algebra_map R K) }
+
+/-- Because `splitting_field_aux` is defined by recursion, we have to make sure all instances
+on `splitting_field_aux` are defined by recursion within the fields. Otherwise, there will be
+instance diamonds such as the following: -/
+example (n : ℕ) {K : Type u} [field K] {f : K[X]} :
+    (add_comm_monoid.nat_module : module ℕ (splitting_field_aux n f)) =
+  @algebra.to_module _ _ _ _ (splitting_field_aux.algebra n _) :=
+rfl
```

Comment:
An alternate home for this could be `test/instance_diamonds`, though I don't care much either way.

Reply:
I'll just leave it here; it's nice to get instant feedback on whether you broke stuff.